File: clients/csharp/Codegame/ServerMessage.cs
namespace AiCup22.Codegame
{
    /// <summary>
    /// Message sent from server
    /// </summary>
    public abstract class ServerMessage
    {
        /// <summary> Write ServerMessage to writer </summary>
        public abstract void WriteTo(System.IO.BinaryWriter writer);

        /// <summary> Read ServerMessage from reader </summary>
        public static ServerMessage ReadFrom(System.IO.BinaryReader reader)
        {
            switch (reader.ReadInt32())
            {
                case UpdateConstants.TAG:
                    return UpdateConstants.ReadFrom(reader);
                case GetOrder.TAG:
                    return GetOrder.ReadFrom(reader);
                case Finish.TAG:
                    return Finish.ReadFrom(reader);
                case DebugUpdate.TAG:
                    return DebugUpdate.ReadFrom(reader);
                default:
                    throw new System.Exception("Unexpected tag value");
            }
        }

        /// <summary>
        /// Update constants
        /// </summary>
        public class UpdateConstants : ServerMessage
        {
            public const int TAG = 0;
        
            /// <summary>
            /// New constants
            /// </summary>
            public AiCup22.Model.Constants Constants { get; set; }
        
            public UpdateConstants() { }
        
            public UpdateConstants(AiCup22.Model.Constants constants)
            {
                this.Constants = constants;
            }
        
            /// <summary> Read UpdateConstants from reader </summary>
            public static new UpdateConstants ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new UpdateConstants();
                result.Constants = AiCup22.Model.Constants.ReadFrom(reader);
                return result;
            }
        
            /// <summary> Write UpdateConstants to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                Constants.WriteTo(writer);
            }
        
            /// <summary> Get string representation of UpdateConstants </summary>
            public override string ToString() {
                string stringResult = "UpdateConstants { ";
                stringResult += "Constants: ";
                stringResult += Constants.ToString();
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// Get order for next tick
        /// </summary>
        public class GetOrder : ServerMessage
        {
            public const int TAG = 1;
        
            /// <summary>
            /// Player's view
            /// </summary>
            public AiCup22.Model.Game PlayerView { get; set; }
            /// <summary>
            /// Whether app is running with debug interface available
            /// </summary>
            public bool DebugAvailable { get; set; }
        
            public GetOrder() { }
        
            public GetOrder(AiCup22.Model.Game playerView, bool debugAvailable)
            {
                this.PlayerView = playerView;
                this.DebugAvailable = debugAvailable;
            }
        
            /// <summary> Read GetOrder from reader </summary>
            public static new GetOrder ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new GetOrder();
                result.PlayerView = AiCup22.Model.Game.ReadFrom(reader);
                result.DebugAvailable = reader.ReadBoolean();
                return result;
            }
        
            /// <summary> Write GetOrder to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                PlayerView.WriteTo(writer);
                writer.Write(DebugAvailable);
            }
        
            /// <summary> Get string representation of GetOrder </summary>
            public override string ToString() {
                string stringResult = "GetOrder { ";
                stringResult += "PlayerView: ";
                stringResult += PlayerView.ToString();
                stringResult += ", ";
                stringResult += "DebugAvailable: ";
                stringResult += DebugAvailable.ToString();
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// Signifies end of the game
        /// </summary>
        public class Finish : ServerMessage
        {
            public const int TAG = 2;
        
        
            public Finish() { }
        
            /// <summary> Read Finish from reader </summary>
            public static new Finish ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new Finish();
                return result;
            }
        
            /// <summary> Write Finish to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
            }
        
            /// <summary> Get string representation of Finish </summary>
            public override string ToString() {
                string stringResult = "Finish { ";
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// Debug update
        /// </summary>
        public class DebugUpdate : ServerMessage
        {
            public const int TAG = 3;
        
        
            public DebugUpdate() { }
        
            /// <summary> Read DebugUpdate from reader </summary>
            public static new DebugUpdate ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new DebugUpdate();
                return result;
            }
        
            /// <summary> Write DebugUpdate to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
            }
        
            /// <summary> Get string representation of DebugUpdate </summary>
            public override string ToString() {
                string stringResult = "DebugUpdate { ";
                stringResult += " }";
                return stringResult;
            }
        }
    }
}

File: clients/fsharp/Codegame/ServerMessage.fs
#nowarn "0058"

namespace AiCup22
namespace AiCup22.Codegame

open AiCup22

/// Update constants
type ServerMessageUpdateConstants = {
    /// New constants
    Constants: Model.Constants;
} with

    /// Write UpdateConstants to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 0
        this.Constants.writeTo writer
        ()

    /// Read UpdateConstants from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Constants = Model.Constants.readFrom reader;
    }

/// Get order for next tick
type ServerMessageGetOrder = {
    /// Player's view
    PlayerView: Model.Game;
    /// Whether app is running with debug interface available
    DebugAvailable: bool;
} with

    /// Write GetOrder to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 1
        this.PlayerView.writeTo writer
        writer.Write this.DebugAvailable
        ()

    /// Read GetOrder from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        PlayerView = Model.Game.readFrom reader;
        DebugAvailable = reader.ReadBoolean()
    }

/// Signifies end of the game
type ServerMessageFinish = struct end with

    /// Write Finish to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 2
        ()

    /// Read Finish from reader
    static member readFrom(reader: System.IO.BinaryReader) = new ServerMessageFinish()

/// Debug update
type ServerMessageDebugUpdate = {
    /// Displayed tick
    DisplayedTick: int;
} with

    /// Write DebugUpdate to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 3
        writer.Write this.DisplayedTick
        ()

    /// Read DebugUpdate from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        DisplayedTick = reader.ReadInt32()
    }

/// Message sent from server
type ServerMessage =
    /// Update constants
    | UpdateConstants of ServerMessageUpdateConstants
    /// Get order for next tick
    | GetOrder of ServerMessageGetOrder
    /// Signifies end of the game
    | Finish of ServerMessageFinish
    /// Debug update
    | DebugUpdate of ServerMessageDebugUpdate
    with

    /// Write ServerMessage to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        match this with
            | UpdateConstants value -> value.writeTo writer
            | GetOrder value -> value.writeTo writer
            | Finish value -> value.writeTo writer
            | DebugUpdate value -> value.writeTo writer

    /// Read ServerMessage from reader
    static member readFrom(reader: System.IO.BinaryReader) =
        match reader.ReadInt32() with
            | 0 -> UpdateConstants (ServerMessageUpdateConstants.readFrom reader)
            | 1 -> GetOrder (ServerMessageGetOrder.readFrom reader)
            | 2 -> Finish (ServerMessageFinish.readFrom reader)
            | 3 -> DebugUpdate (ServerMessageDebugUpdate.readFrom reader)
            | x -> failwith (sprintf "Unexpected tag %d" x)